File: GenesisAndroid/src/test/java/com/emerchantpay/gateway/genesisandroid/internal/ResponseUnitTest.java
package com.emerchantpay.gateway.genesisandroid.internal;

import com.emerchantpay.gateway.genesisandroid.api.internal.response.Response;
import com.emerchantpay.gateway.genesisandroid.api.models.GenesisError;
import com.emerchantpay.gateway.genesisandroid.api.models.Transaction;

import org.junit.Before;
import org.junit.Test;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ResponseUnitTest {
    private Response response;
    private Transaction transaction;
    private GenesisError error;

    @Before
    public void setup() {
        response = mock(Response.class);
        transaction = mock(Transaction.class);
    }

    @Test
    public void testFailedResponse() {
        when(response.getTransaction()).thenReturn(transaction);
        when(response.getStatus()).thenReturn("error");

        // Init Genesis error
        error = new GenesisError(response.getStatus());

        when(response.isSuccess()).thenReturn(false);
        when(response.getError()).thenReturn(error);

        assertFalse(response.isSuccess());
        assertEquals(response.getError().getMessage(), "error");
    }

    @Test
    public void testSuccessResponse() {
        when(response.getTransaction()).thenReturn(transaction);
        when(response.getStatus()).thenReturn("approved");

        // Init Genesis error
        error = new GenesisError(response.getStatus());

        when(response.isSuccess()).thenReturn(true);
        when(response.getError()).thenReturn(error);

        assertTrue(response.isSuccess());
        assertEquals(response.getError().getMessage(), "approved");
    }
}

File: GenesisAndroid/src/test/java/com/emerchantpay/gateway/genesisandroid/internal/ResponseUnitTest.kt
package com.emerchantpay.gateway.genesisandroid.internal

import com.emerchantpay.gateway.genesisandroid.api.internal.response.Response
import com.emerchantpay.gateway.genesisandroid.api.models.GenesisError
import com.emerchantpay.gateway.genesisandroid.api.models.Transaction
import junit.framework.Assert.*
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock

class ResponseUnitTest {
    private var response: Response? = null
    private var transaction: Transaction? = null
    private var error: GenesisError? = null

    @Before
    fun setup() {
        response = mock(Response::class.java)
        transaction = mock(Transaction::class.java)
    }

    @Test
    fun testFailedResponse() {
        `when`(response!!.transaction).thenReturn(transaction)
        `when`(response!!.status).thenReturn("error")

        // Init Genesis error
        error = response!!.status?.let { GenesisError(it) }

        `when`(response!!.isSuccess).thenReturn(false)
        `when`(response!!.error).thenReturn(error)

        assertFalse(response!!.isSuccess!!)
        assertEquals(response!!.error!!.message, "error")
    }

    @Test
    fun testSuccessResponse() {
        `when`(response!!.transaction).thenReturn(transaction)
        `when`(response!!.status).thenReturn("approved")

        // Init Genesis error
        error = response!!.status?.let { GenesisError(it) }

        `when`(response!!.isSuccess).thenReturn(true)
        `when`(response!!.error).thenReturn(error)

        assertTrue(response!!.isSuccess!!)
        assertEquals(response!!.error!!.message, "approved")
    }
}
